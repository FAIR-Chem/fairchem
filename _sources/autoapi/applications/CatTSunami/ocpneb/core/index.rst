:py:mod:`applications.CatTSunami.ocpneb.core`
=============================================

.. py:module:: applications.CatTSunami.ocpneb.core


Submodules
----------
.. toctree::
   :titlesonly:
   :maxdepth: 1

   autoframe/index.rst
   ocpneb/index.rst
   reaction/index.rst


