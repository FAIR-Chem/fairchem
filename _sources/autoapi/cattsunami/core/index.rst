cattsunami.core
===============

.. py:module:: cattsunami.core


Submodules
----------

.. toctree::
   :maxdepth: 1

   /autoapi/cattsunami/core/autoframe/index
   /autoapi/cattsunami/core/ocpneb/index
   /autoapi/cattsunami/core/reaction/index


Classes
-------

.. autoapisummary::

   cattsunami.core.OCPNEB
   cattsunami.core.Reaction


Package Contents
----------------

.. py:class:: OCPNEB(images, checkpoint_path, k=0.1, fmax=0.05, climb=False, parallel=False, remove_rotation_and_translation=False, world=None, dynamic_relaxation=True, scale_fmax=0.0, method='aseneb', allow_shared_calculator=False, precon=None, cpu=False, batch_size=4)

   Bases: :py:obj:`ase.neb.DyNEB`


   .. py:attribute:: batch_size


   .. py:attribute:: config


   .. py:attribute:: trainer


   .. py:attribute:: a2g


   .. py:attribute:: intermediate_energies
      :value: []



   .. py:attribute:: intermediate_forces
      :value: []



   .. py:attribute:: cached
      :value: False



   .. py:method:: load_checkpoint(checkpoint_path: str) -> None

      Load existing trained model

      :param checkpoint_path: string
                              Path to trained model



   .. py:method:: get_forces()

      Evaluate and return the forces.



   .. py:method:: set_positions(positions)


   .. py:method:: get_precon_forces(forces, energies, images)


.. py:class:: Reaction(reaction_db_path: str, adsorbate_db_path: str, reaction_id_from_db: int | None = None, reaction_str_from_db: str | None = None, reaction_type: str | None = None)

   Initialize Reaction object


   .. py:attribute:: reaction_db_path


   .. py:attribute:: reaction_db


   .. py:attribute:: adsorbate_db


   .. py:attribute:: entry


   .. py:method:: get_desorption_mapping(reactant)

      Get mapping for desorption reaction



