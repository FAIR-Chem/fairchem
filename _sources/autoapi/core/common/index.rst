:py:mod:`core.common`
=====================

.. py:module:: core.common

.. autoapi-nested-parse::

   Copyright (c) Meta, Inc. and its affiliates.

   This source code is licensed under the MIT license found in the
   LICENSE file in the root directory of this source tree.



Subpackages
-----------
.. toctree::
   :titlesonly:
   :maxdepth: 3

   relaxation/index.rst


Submodules
----------
.. toctree::
   :titlesonly:
   :maxdepth: 1

   data_parallel/index.rst
   distutils/index.rst
   flags/index.rst
   gp_utils/index.rst
   hpo_utils/index.rst
   logger/index.rst
   registry/index.rst
   transforms/index.rst
   tutorial_utils/index.rst
   typing/index.rst
   utils/index.rst


