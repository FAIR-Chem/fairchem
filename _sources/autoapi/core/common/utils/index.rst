core.common.utils
=================

.. py:module:: core.common.utils

.. autoapi-nested-parse::

   Copyright (c) Meta, Inc. and its affiliates.

   This source code is licensed under the MIT license found in the
   LICENSE file in the root directory of this source tree.



Classes
-------

.. autoapisummary::

   core.common.utils.UniqueKeyLoader
   core.common.utils.Complete
   core.common.utils.SeverityLevelBetween


Functions
---------

.. autoapisummary::

   core.common.utils.pyg2_data_transform
   core.common.utils.save_checkpoint
   core.common.utils.warmup_lr_lambda
   core.common.utils.print_cuda_usage
   core.common.utils.conditional_grad
   core.common.utils.plot_histogram
   core.common.utils.collate
   core.common.utils.add_edge_distance_to_graph
   core.common.utils._import_local_file
   core.common.utils.setup_experimental_imports
   core.common.utils._get_project_root
   core.common.utils.setup_imports
   core.common.utils.dict_set_recursively
   core.common.utils.parse_value
   core.common.utils.create_dict_from_args
   core.common.utils.load_config
   core.common.utils.build_config
   core.common.utils.create_grid
   core.common.utils.save_experiment_log
   core.common.utils.get_pbc_distances
   core.common.utils.radius_graph_pbc
   core.common.utils.get_max_neighbors_mask
   core.common.utils.get_pruned_edge_idx
   core.common.utils.merge_dicts
   core.common.utils.setup_logging
   core.common.utils.compute_neighbors
   core.common.utils.check_traj_files
   core.common.utils.new_trainer_context
   core.common.utils._resolve_scale_factor_submodule
   core.common.utils._report_incompat_keys
   core.common.utils.load_state_dict
   core.common.utils.scatter_det
   core.common.utils.get_commit_hash
   core.common.utils.cg_change_mat
   core.common.utils.irreps_sum
   core.common.utils.update_config
   core.common.utils.get_loss_module


Module Contents
---------------

.. py:class:: UniqueKeyLoader(stream)

   Bases: :py:obj:`yaml.SafeLoader`


   .. py:method:: construct_mapping(node, deep=False)


.. py:function:: pyg2_data_transform(data: torch_geometric.data.Data)

   if we're on the new pyg (2.0 or later) and if the Data stored is in older format
   we need to convert the data to the new format


.. py:function:: save_checkpoint(state, checkpoint_dir: str = 'checkpoints/', checkpoint_file: str = 'checkpoint.pt') -> str

.. py:class:: Complete

   .. py:method:: __call__(data)


.. py:function:: warmup_lr_lambda(current_step: int, optim_config)

   Returns a learning rate multiplier.
   Till `warmup_steps`, learning rate linearly increases to `initial_lr`,
   and then gets multiplied by `lr_gamma` every time a milestone is crossed.


.. py:function:: print_cuda_usage() -> None

.. py:function:: conditional_grad(dec)

   Decorator to enable/disable grad depending on whether force/energy predictions are being made


.. py:function:: plot_histogram(data, xlabel: str = '', ylabel: str = '', title: str = '')

.. py:function:: collate(data_list)

.. py:function:: add_edge_distance_to_graph(batch, device='cpu', dmin: float = 0.0, dmax: float = 6.0, num_gaussians: int = 50)

.. py:function:: _import_local_file(path: pathlib.Path, *, project_root: pathlib.Path) -> None

   Imports a Python file as a module

   :param path: The path to the file to import
   :type path: Path
   :param project_root: The root directory of the project (i.e., the "ocp" folder)
   :type project_root: Path


.. py:function:: setup_experimental_imports(project_root: pathlib.Path) -> None

   Import selected directories of modules from the "experimental" subdirectory.

   If a file named ".include" is present in the "experimental" subdirectory,
   this will be read as a list of experimental subdirectories whose module
   (including in any subsubdirectories) should be imported.

   :param project_root: The root directory of the project (i.e., the "ocp" folder)


.. py:function:: _get_project_root() -> pathlib.Path

   Gets the root folder of the project (the "ocp" folder)
   :return: The absolute path to the project root.


.. py:function:: setup_imports(config: dict | None = None) -> None

.. py:function:: dict_set_recursively(dictionary, key_sequence, val) -> None

.. py:function:: parse_value(value)

   Parse string as Python literal if possible and fallback to string.


.. py:function:: create_dict_from_args(args: list, sep: str = '.')

   Create a (nested) dictionary from console arguments.
   Keys in different dictionary levels are separated by sep.


.. py:function:: load_config(path: str, previous_includes: list | None = None)

.. py:function:: build_config(args, args_override)

.. py:function:: create_grid(base_config, sweep_file: str)

.. py:function:: save_experiment_log(args, jobs, configs)

.. py:function:: get_pbc_distances(pos, edge_index, cell, cell_offsets, neighbors, return_offsets: bool = False, return_distance_vec: bool = False)

.. py:function:: radius_graph_pbc(data, radius, max_num_neighbors_threshold, enforce_max_neighbors_strictly: bool = False, pbc=None)

.. py:function:: get_max_neighbors_mask(natoms, index, atom_distance, max_num_neighbors_threshold, degeneracy_tolerance: float = 0.01, enforce_max_strictly: bool = False)

   Give a mask that filters out edges so that each atom has at most
   `max_num_neighbors_threshold` neighbors.
   Assumes that `index` is sorted.

   Enforcing the max strictly can force the arbitrary choice between
   degenerate edges. This can lead to undesired behaviors; for
   example, bulk formation energies which are not invariant to
   unit cell choice.

   A degeneracy tolerance can help prevent sudden changes in edge
   existence from small changes in atom position, for example,
   rounding errors, slab relaxation, temperature, etc.


.. py:function:: get_pruned_edge_idx(edge_index, num_atoms: int, max_neigh: float = 1000000000.0) -> torch.Tensor

.. py:function:: merge_dicts(dict1: dict, dict2: dict)

   Recursively merge two dictionaries.
   Values in dict2 override values in dict1. If dict1 and dict2 contain a dictionary as a
   value, this will call itself recursively to merge these dictionaries.
   This does not modify the input dictionaries (creates an internal copy).
   Additionally returns a list of detected duplicates.
   Adapted from https://github.com/TUM-DAML/seml/blob/master/seml/utils.py

   :param dict1: First dict.
   :type dict1: dict
   :param dict2: Second dict. Values in dict2 will override values from dict1 in case they share the same key.
   :type dict2: dict

   :returns: **return_dict** -- Merged dictionaries.
   :rtype: dict


.. py:class:: SeverityLevelBetween(min_level: int, max_level: int)

   Bases: :py:obj:`logging.Filter`


   Filter instances are used to perform arbitrary filtering of LogRecords.

   Loggers and Handlers can optionally use Filter instances to filter
   records as desired. The base filter class only allows events which are
   below a certain point in the logger hierarchy. For example, a filter
   initialized with "A.B" will allow events logged by loggers "A.B",
   "A.B.C", "A.B.C.D", "A.B.D" etc. but not "A.BB", "B.A.B" etc. If
   initialized with the empty string, all events are passed.


   .. py:method:: filter(record) -> bool

      Determine if the specified record is to be logged.

      Returns True if the record should be logged, or False otherwise.
      If deemed appropriate, the record may be modified in-place.



.. py:function:: setup_logging() -> None

.. py:function:: compute_neighbors(data, edge_index)

.. py:function:: check_traj_files(batch, traj_dir) -> bool

.. py:function:: new_trainer_context(*, config: dict[str, Any], distributed: bool = False)

.. py:function:: _resolve_scale_factor_submodule(model: torch.nn.Module, name: str)

.. py:function:: _report_incompat_keys(model: torch.nn.Module, keys: torch.nn.modules.module._IncompatibleKeys, strict: bool = False) -> tuple[list[str], list[str]]

.. py:function:: load_state_dict(module: torch.nn.Module, state_dict: collections.abc.Mapping[str, torch.Tensor], strict: bool = True) -> tuple[list[str], list[str]]

.. py:function:: scatter_det(*args, **kwargs)

.. py:function:: get_commit_hash()

.. py:function:: cg_change_mat(ang_mom: int, device: str = 'cpu') -> torch.tensor

.. py:function:: irreps_sum(ang_mom: int) -> int

   Returns the sum of the dimensions of the irreps up to the specified angular momentum.

   :param ang_mom: max angular momenttum to sum up dimensions of irreps


.. py:function:: update_config(base_config)

   Configs created prior to FAIRChem/OCP 2.0 are organized a little different than they
   are now. Update old configs to fit the new expected structure.


.. py:function:: get_loss_module(loss_name)

