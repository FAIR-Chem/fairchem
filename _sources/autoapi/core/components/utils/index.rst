core.components.utils
=====================

.. py:module:: core.components.utils

.. autoapi-nested-parse::

   Copyright (c) Meta, Inc. and its affiliates.

   This source code is licensed under the MIT license found in the
   LICENSE file in the root directory of this source tree.



Classes
-------

.. autoapisummary::

   core.components.utils.ManagedAttribute


Module Contents
---------------

.. py:class:: ManagedAttribute(enforced_type: type | None = None)

   A descriptor helper to manage setting/access to an attribute of a class


   .. py:attribute:: _enforced_type


   .. py:method:: __set_name__(owner: fairchem.core.components.runner.Runner | fairchem.core.components.reducer.Reducer, name: str)


   .. py:method:: __get__(obj: fairchem.core.components.runner.Runner | fairchem.core.components.reducer.Reducer, objtype: type[fairchem.core.components.runner.Runner | fairchem.core.components.reducer.Reducer] | None = None)


   .. py:method:: __set__(obj: fairchem.core.components.runner.Runner | fairchem.core.components.reducer.Reducer, value: Any)


