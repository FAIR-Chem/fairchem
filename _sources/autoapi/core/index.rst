:py:mod:`core`
==============

.. py:module:: core

.. autoapi-nested-parse::

   Copyright (c) Meta, Inc. and its affiliates.

   This source code is licensed under the MIT license found in the
   LICENSE file in the root directory of this source tree.



Subpackages
-----------
.. toctree::
   :titlesonly:
   :maxdepth: 3

   common/index.rst
   datasets/index.rst
   models/index.rst
   modules/index.rst
   preprocessing/index.rst
   scripts/index.rst
   tasks/index.rst
   trainers/index.rst


Submodules
----------
.. toctree::
   :titlesonly:
   :maxdepth: 1

   _cli/index.rst


