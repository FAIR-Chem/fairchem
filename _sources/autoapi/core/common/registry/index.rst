:py:mod:`core.common.registry`
==============================

.. py:module:: core.common.registry

.. autoapi-nested-parse::

   Copyright (c) Meta, Inc. and its affiliates.

   This source code is licensed under the MIT license found in the
   LICENSE file in the root directory of this source tree.

   # Copyright (c) Meta, Inc. and its affiliates.
   # Borrowed from https://github.com/facebookresearch/pythia/blob/master/pythia/common/registry.py.

   Registry is central source of truth. Inspired from Redux's concept of
   global store, Registry maintains mappings of various information to unique
   keys. Special functions in registry can be used as decorators to register
   different kind of classes.

   Import the global registry object using

   ``from fairchem.core.common.registry import registry``

   Various decorators for registry different kind of classes with unique keys

   - Register a model: ``@registry.register_model``



Module Contents
---------------

Classes
~~~~~~~

.. autoapisummary::

   core.common.registry.Registry



Functions
~~~~~~~~~

.. autoapisummary::

   core.common.registry._get_absolute_mapping



Attributes
~~~~~~~~~~

.. autoapisummary::

   core.common.registry.R
   core.common.registry.NestedDict
   core.common.registry.registry


.. py:data:: R

   

.. py:data:: NestedDict

   

.. py:function:: _get_absolute_mapping(name: str)


.. py:class:: Registry


   Class for registry object which acts as central source of truth.

   .. py:attribute:: mapping
      :type: ClassVar[NestedDict]

      

   .. py:method:: register_task(name: str)
      :classmethod:

      Register a new task to registry with key 'name'
      :param name: Key with which the task will be registered.

      Usage::
          from fairchem.core.common.registry import registry
          from fairchem.core.tasks import BaseTask
          @registry.register_task("train")
          class TrainTask(BaseTask):
              ...


   .. py:method:: register_dataset(name: str)
      :classmethod:

      Register a dataset to registry with key 'name'

      :param name: Key with which the dataset will be registered.

      Usage::

          from fairchem.core.common.registry import registry
          from fairchem.core.datasets import BaseDataset

          @registry.register_dataset("qm9")
          class QM9(BaseDataset):
              ...


   .. py:method:: register_model(name: str)
      :classmethod:

      Register a model to registry with key 'name'

      :param name: Key with which the model will be registered.

      Usage::

          from fairchem.core.common.registry import registry
          from fairchem.core.modules.layers import CGCNNConv

          @registry.register_model("cgcnn")
          class CGCNN():
              ...


   .. py:method:: register_logger(name: str)
      :classmethod:

      Register a logger to registry with key 'name'

      :param name: Key with which the logger will be registered.

      Usage::

          from fairchem.core.common.registry import registry

          @registry.register_logger("wandb")
          class WandBLogger():
              ...


   .. py:method:: register_trainer(name: str)
      :classmethod:

      Register a trainer to registry with key 'name'

      :param name: Key with which the trainer will be registered.

      Usage::

          from fairchem.core.common.registry import registry

          @registry.register_trainer("active_discovery")
          class ActiveDiscoveryTrainer():
              ...


   .. py:method:: register(name: str, obj) -> None
      :classmethod:

      Register an item to registry with key 'name'

      :param name: Key with which the item will be registered.

      Usage::

          from fairchem.core.common.registry import registry

          registry.register("config", {})


   .. py:method:: __import_error(name: str, mapping_name: str) -> RuntimeError
      :classmethod:


   .. py:method:: get_class(name: str, mapping_name: str)
      :classmethod:


   .. py:method:: get_task_class(name: str)
      :classmethod:


   .. py:method:: get_dataset_class(name: str)
      :classmethod:


   .. py:method:: get_model_class(name: str)
      :classmethod:


   .. py:method:: get_logger_class(name: str)
      :classmethod:


   .. py:method:: get_trainer_class(name: str)
      :classmethod:


   .. py:method:: get(name: str, default=None, no_warning: bool = False)
      :classmethod:

      Get an item from registry with key 'name'

      :param name: Key whose value needs to be retrieved.
      :type name: string
      :param default: If passed and key is not in registry, default value will
                      be returned with a warning. Default: None
      :param no_warning: If passed as True, warning when key doesn't exist
                         will not be generated. Useful for cgcnn's
                         internal operations. Default: False
      :type no_warning: bool

      Usage::

          from fairchem.core.common.registry import registry

          config = registry.get("config")


   .. py:method:: unregister(name: str)
      :classmethod:

      Remove an item from registry with key 'name'

      :param name: Key which needs to be removed.

      Usage::

          from fairchem.core.common.registry import registry

          config = registry.unregister("config")



.. py:data:: registry

   

