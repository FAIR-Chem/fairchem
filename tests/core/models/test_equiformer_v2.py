"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
"""

from __future__ import annotations

import copy
import io
import os

import pytest
import requests
import torch
from ase.io import read
from torch.nn.parallel.distributed import DistributedDataParallel

from fairchem.core.common.registry import registry
from fairchem.core.common.test_utils import PGConfig, spawn_multi_process
from fairchem.core.common.utils import load_state_dict, setup_imports
from fairchem.core.datasets import data_list_collater
from fairchem.core.models.equiformer_v2.so3 import (
    CoefficientMappingModule,
    SO3_Embedding,
)
from fairchem.core.preprocessing import AtomsToGraphs


@pytest.fixture(scope="class")
def load_data(request):
    atoms = read(
        os.path.join(os.path.dirname(os.path.abspath(__file__)), "atoms.json"),
        index=0,
        format="json",
    )
    a2g = AtomsToGraphs(
        max_neigh=200,
        radius=6,
        r_edges=False,
        r_fixed=True,
    )
    data_list = a2g.convert_all([atoms])
    request.cls.data = data_list[0]


def _load_model():
    torch.manual_seed(4)
    setup_imports()

    # download and load weights.
    checkpoint_url = "https://dl.fbaipublicfiles.com/opencatalystproject/models/2023_06/oc20/s2ef/eq2_31M_ec4_allmd.pt"

    # load buffer into memory as a stream
    # and then load it with torch.load
    r = requests.get(checkpoint_url, stream=True)
    r.raise_for_status()
    checkpoint = torch.load(io.BytesIO(r.content), map_location=torch.device("cpu"))

    model = registry.get_model_class("equiformer_v2")(
        None,
        -1,
        1,
        use_pbc=True,
        regress_forces=True,
        otf_graph=True,
        max_neighbors=20,
        max_radius=12.0,
        max_num_elements=90,
        num_layers=8,
        sphere_channels=128,
        attn_hidden_channels=64,
        num_heads=8,
        attn_alpha_channels=64,
        attn_value_channels=16,
        ffn_hidden_channels=128,
        norm_type="layer_norm_sh",
        lmax_list=[4],
        mmax_list=[2],
        grid_resolution=18,
        num_sphere_samples=128,
        edge_channels=128,
        use_atom_edge_embedding=True,
        distance_function="gaussian",
        num_distance_basis=512,
        attn_activation="silu",
        use_s2_act_attn=False,
        ffn_activation="silu",
        use_gate_act=False,
        use_grid_mlp=True,
        alpha_drop=0.1,
        drop_path_rate=0.1,
        proj_drop=0.0,
        weight_init="uniform",
    )

    new_dict = {k[len("module.") * 2 :]: v for k, v in checkpoint["state_dict"].items()}
    load_state_dict(model, new_dict)

    # Precision errors between mac vs. linux compound with multiple layers,
    # so we explicitly set the number of layers to 1 (instead of all 8).
    # The other alternative is to have different snapshots for mac vs. linux.
    model.num_layers = 1
    return model


@pytest.fixture(scope="class")
def load_model(request):
    request.cls.model = _load_model()


def _runner(data):
    # serializing the model through python multiprocess results in precision errors, so we get a fresh model here
    model = _load_model()
    ddp_model = DistributedDataParallel(model)
    outputs = ddp_model(data_list_collater([data]))
    return {k: v.detach() for k, v in outputs.items()}


@pytest.mark.usefixtures("load_data")
@pytest.mark.usefixtures("load_model")
class TestEquiformerV2:
    def test_energy_force_shape(self, snapshot):
        # Recreate the Data object to only keep the necessary features.
        data = self.data
        model = copy.deepcopy(self.model)

        # Pass it through the model.
        outputs = model(data_list_collater([data]))
        print(outputs)
        energy, forces = outputs["energy"], outputs["forces"]

        assert snapshot == energy.shape
        assert snapshot == pytest.approx(energy.detach())

        assert snapshot == forces.shape
        assert snapshot == pytest.approx(forces.detach().mean(0))

    def test_ddp(self, snapshot):
        data_dist = self.data.clone().detach()
        config = PGConfig(backend="gloo", world_size=1, gp_group_size=1, use_gp=False)
        output = spawn_multi_process(config, _runner, data_dist)
        assert len(output) == 1
        energy, forces = output[0]["energy"], output[0]["forces"]
        assert snapshot == energy.shape
        assert snapshot == pytest.approx(energy.detach())
        assert snapshot == forces.shape
        assert snapshot == pytest.approx(forces.detach().mean(0))

    def test_gp(self, snapshot):
        data_dist = self.data.clone().detach()
        config = PGConfig(backend="gloo", world_size=2, gp_group_size=2, use_gp=True)
        output = spawn_multi_process(config, _runner, data_dist)
        assert len(output) == 2
        energy, forces = output[0]["energy"], output[0]["forces"]
        assert snapshot == energy.shape
        assert snapshot == pytest.approx(energy.detach())
        assert snapshot == forces.shape
        assert snapshot == pytest.approx(forces.detach().mean(0))


class TestMPrimaryLPrimary:
    def test_mprimary_lprimary_mappings(self):
        def sign(x):
            return 1 if x >= 0 else -1

        device = torch.device("cpu")
        lmax_list = [6, 8]
        mmax_list = [3, 6]
        for lmax in lmax_list:
            for mmax in mmax_list:
                c = CoefficientMappingModule([lmax], [mmax])

                embedding = SO3_Embedding(
                    length=1,
                    lmax_list=[lmax],
                    num_channels=1,
                    device=device,
                    dtype=torch.float32,
                )

                """
                Generate L_primary matrix
                L0: 0.00 ~ L0M0
                L1: -1.01 1.00 1.01 ~ L1M(-1),L1M0,L1M1
                L2: -2.02 -2.01 2.00 2.01 2.02 ~ L2M(-2),L2M(-1),L2M0,L2M1,L2M2
                """
                test_matrix_lp = []
                for l in range(lmax + 1):
                    max_m = min(l, mmax)
                    for m in range(-max_m, max_m + 1):
                        v = l * sign(m) + 0.01 * m  # +/- l . 00 m
                        test_matrix_lp.append(v)

                test_matrix_lp = (
                    torch.tensor(test_matrix_lp).reshape(1, -1, 1).to(torch.float32)
                )

                """
                Generate M_primary matrix
                M0: 0.00 , 1.00, 2.00, ... , LMax ~ M0L0, M0L1, .., M0L(LMax)
                M1: 1.01, 2.01, .., LMax.01, -1.01, -2.01, -LMax.01 ~ L1M1, L2M1, .., L(LMax)M1, L1M(-1), L2M(-1), ... , L(LMax)M(-1)
                """
                test_matrix_mp = []
                for m in range(max_m + 1):
                    for l in range(m, lmax + 1):
                        v = l + 0.01 * m  # +/- l . 00 m
                        test_matrix_mp.append(v)
                    if m > 0:
                        for l in range(m, lmax + 1):
                            v = -(l + 0.01 * m)  # +/- l . 00 m
                            test_matrix_mp.append(v)

                test_matrix_mp = (
                    torch.tensor(test_matrix_mp).reshape(1, -1, 1).to(torch.float32)
                )

                embedding.embedding = test_matrix_lp.clone()

                embedding._m_primary(c)
                mp = embedding.embedding.clone()
                (test_matrix_mp == mp).all()

                embedding._l_primary(c)
                lp = embedding.embedding.clone()
                (test_matrix_lp == lp).all()

