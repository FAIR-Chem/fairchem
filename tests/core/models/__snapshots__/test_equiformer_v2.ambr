# serializer version: 1
# name: TestEquiformerV2.test_ddp
  Size(
    1,
  )
# ---
# name: TestEquiformerV2.test_ddp.1
  Approx(
  	array([0.12408741], dtype=float32), 
  	rtol=0.001, 
  	atol=0.001
  )
# ---
# name: TestEquiformerV2.test_ddp.2
  Size(
    34,
    3,
  )
# ---
# name: TestEquiformerV2.test_ddp.3
  Approx(
  	array([ 1.4928594e-03, -7.4167736e-05,  2.9909366e-03], dtype=float32), 
  	rtol=0.001, 
  	atol=0.001
  )
# ---
# name: TestEquiformerV2.test_energy_force_shape
  Size(
    1,
  )
# ---
# name: TestEquiformerV2.test_energy_force_shape.1
  Approx(
  	array([0.12408741], dtype=float32), 
  	rtol=0.001, 
  	atol=0.001
  )
# ---
# name: TestEquiformerV2.test_energy_force_shape.2
  Size(
    34,
    3,
  )
# ---
# name: TestEquiformerV2.test_energy_force_shape.3
  Approx(
  	array([ 1.4928594e-03, -7.4167736e-05,  2.9909366e-03], dtype=float32), 
  	rtol=0.001, 
  	atol=0.001
  )
# ---
# name: TestEquiformerV2.test_gp
  Size(
    1,
  )
# ---
# name: TestEquiformerV2.test_gp.1
  Approx(
  	array([0.12408741], dtype=float32), 
  	rtol=0.001, 
  	atol=0.001
  )
# ---
# name: TestEquiformerV2.test_gp.2
  Size(
    34,
    3,
  )
# ---
# name: TestEquiformerV2.test_gp.3
  Approx(
  	array([ 1.4928658e-03, -7.4134972e-05,  2.9909210e-03], dtype=float32), 
  	rtol=0.001, 
  	atol=0.001
  )
# ---
